### ISSUE TITLE
submariner should accept "Calico" (with capital C) as network plugin

### ISSUE THREAD

**Issue by dberg1** [submariner-io/submariner#1652]
My OpenShift cluster is configured to use Calico as CNI network plugin instead of the default OpenShiftSDN:
```
apiVersion: config.openshift.io/v1
kind: Network
metadata:
  name: cluster
spec:
  clusterNetwork:
    - cidr: 172.17.0.0/18
      hostPrefix: 23
  externalIP:
    policy: {}
  networkType: Calico
  serviceNetwork:
    - 172.21.0.0/16
```
unfortunately, submariner does not recognize "Calico" as a supported CNI network plugin because of the capital C:
```
# subctl diagnose all
...
 • Checking Submariner support for the CNI network plugin  ...
 ✗ Checking Submariner support for the CNI network plugin
 ✗ The detected CNI network plugin ("Calico") is not supported by Submariner. Supported network plugins: [generic canal-flannel weave-net OpenShiftSDN OVNKubernetes calico]
```

**What happened**:
`submariner-route-agent` does not create the required iptables rules because it does not recognize the "Calico" network plugin.

**What you expected to happen**:
submariner should support "Calico" as this is the suggested name to use in the calico documentation:
https://projectcalico.docs.tigera.io/getting-started/openshift/installation

> $ sed -i 's/OpenShiftSDN/Calico/' install-config.yaml

**How to reproduce it (as minimally and precisely as possible)**:
Create an OpenShift cluster using "Calico" as network plugin.

**Anything else we need to know?**:

**Environment**:

**Comment by dberg1** [comment]
Changing pkg/routeagent_driver/constants/constants.go:
```
NetworkPluginCalico        = "calico"
```
to
```
NetworkPluginCalico        = "Calico"
```
seems to fix the issue for me (but you might want to accept both "calico" and "Calico").
At least the `submariner-route-agent` now creates the iptables rules properly.

**Comment by sridhargaddam** [comment]
> seems to fix the issue for me (but you might want to accept both "calico" and "Calico"). At least the `submariner-route-agent` now creates the iptables rules properly.

Thanks for reporting the issue @dberg1. Can you please confirm that after you modified the name to "Calico" the connectivity between the clusters is working fine?

**Comment by dberg1** [comment]
My setup is a bit special as I am trying to connect an OpenShift cluster using OpenShiftSDN to another OpenShift cluster using Calico.
The IPsec connection is successfully established between the submariner gateway nodes on both clusters (`oc -n submariner-operator describe Gateway` says everything is ok on both clusters).
I can access services exported by the Calico cluster from the OpenShiftSDN cluster just fine.
However I still have issues accessing services exported by the OpenShiftSDN cluster from the Calico cluster. It only works if the pod that is trying to access the service is scheduled on the submariner gateway node itself.
Must be something with the submariner VXLAN tunnel (`vx-submariner`) on the Calico cluster.
I created the IPPools on the Calico cluster to not perform NAT on the OpenShiftSDN cluster service IPs.
But I don't think my problem is related to this Git issue.

**Comment by sridhargaddam** [comment]
> But I don't think my problem is related to this Git issue.

Agree, its not related to this issue. Please report a separate one if required. 
PS: You can try running `subctl diagnose firewall intra-cluster ....` on the OCP Calico cluster and it might give some pointers.

**Comment by dberg1** [comment]
@nyechiel Let me know if you want to take a look at our ROKS cluster.
`subctl diagnose firewall intra-cluster` reports no issue but connection from nodes other than the submariner node to services exported by the remote cluster still doesn't work.

**Comment by dberg1** [comment]
Here is my issue with submariner + ROKS.
Cluster B is using Calico.
Node 2 in cluster B tries to connect to a service exported by cluster A.
The SYN packet is sent to the submariner gateway node through the submariner VXLAN tunnel.
The submariner gateway node forwards the SYN packet through the IPsec tunnel established with remote cluster A.
Cluster A sends the SYN+ACK response through the IPsec tunnel.
The submariner gateway node on cluster B forwards the SYN+ACK to node 2 through the eth0 network interface.
Node 2 never receives the SYN+ACK response packet (tcpdump shows nothing).
It looks like the SYN+ACK packet is discarded by some sort of firewall between the submariner gateway node and node 2.

```
  Cluster A                           Cluster B                   
+-----------+        IPsec          +-----------+
|  GW node  |=======================|  GW node  |
|           |<--------SYN-----------|           |
|           |-------SYN+ACK-------->|           |
+-----------+                       +-----------+
                                        ^   |
                                    SYN |   | SYN+ACK
                        (vx-submariner) |   v (eth0)
                                        |   X
                                        |
                                    +-----------+
                                    |   node2   |
                                    |           |
                                    |           |
                                    +-----------+
```

**Comment by sridhargaddam** [comment]
@dberg1 Thanks for the analysis. As you might have seen, there is asymmetric routing involved on the `ClusterB GW node` (i.e, incoming packets come on vx-submariner but leave from some other interface like eth0 in this example). To address this kind of situations, we can tweak reverse_path_filtering settings. 

https://sysctl-explorer.net/net/ipv4/rp_filter/

Please try configuring rp_filter setting to 2 manually on the Gateway node of Cluster B (as a reference you can see this - https://github.com/submariner-io/submariner/blob/1324572a65c049282eb3c49003bdc00344978aa9/pkg/netlink/netlink.go#L141)

**Comment by dberg1** [comment]
`net.ipv4.conf.vx-submariner.rp_filter` is already set to `2` on submariner gateway node.
Tried both:
```
# chroot /host sysctl net.ipv4.conf.eth0.rp_filter=2
# chroot /host sysctl net.ipv4.conf.all.rp_filter=2
```
on submariner gateway node (and even on node 2), but it does not seem to help.

By the way, I also tried to disable TCP segmentation offload on `eth0` as it was enabled:
```
# ethtool -k eth0
Features for eth0:
...
tcp-segmentation-offload: on
        tx-tcp-segmentation: on
        tx-tcp-ecn-segmentation: on
        tx-tcp6-segmentation: on
        tx-tcp-mangleid-segmentation: off
```
but it did not help either.
The strange thing is that I can see the SYN+ACK packet being forwarded to node 2 on `eth0` on the submariner gateway node (using `tcpdump`) with a correct destination MAC address, but I don't see it on `eth0` on node 2. Even if the packet was filtered by `iptables`, I should see it at the network interface level before it reaches the network stack.

**Comment by sridhargaddam** [comment]
My guess would be that SYN+ACK packet from `ClusterB GW node` has sourceIP as ClusterIP Svc in Node A. Since the sourceIP does not match any of the interface IPs on ClusterB GW node, the underlying infrastructure might be dropping the packets silently as part of Anti-spoofing rules (i.e., it might only allow traffic that has the original IP and MAC assigned to the `ClusterB Gw node`).

**Comment by sridhargaddam** [comment]
@dberg1 Yossi is working on fixing the "Calico" mismatch as originally reported in this issue. Can you please open a new issue to discuss/track the problem you are facing.

**Comment by dberg1** [comment]
Opened https://github.com/submariner-io/submariner/issues/1670 to track the submariner+ROKS issues.

**Comment by yboaron** [comment]
@dberg1, If you can still access your Ocp cluster with calico it will be useful if you can share the following:
- NetworkPlugin env var from Routeagent daemonset manifest [1]
- logs from one of submariner-routeagent pods [2]

it may hint at the source of the problem

[1] kubectl get ds -n submariner-operator submariner-routeagent -o yaml | grep -A 2 NETWORKPLU

[2] kubectl logs -n submariner-operator submariner-routeagent-j8jf4

**Comment by dberg1** [comment]
`SUBMARINER_NETWORKPLUGIN=Calico`

Logs:
```
+ trap 'exit 1' SIGTERM SIGINT
+ SUBMARINER_VERBOSITY=2
+ '[' false == true ']'
+ DEBUG=-v=2
+ exec submariner-route-agent -v=2 -alsologtostderr
I0124 15:50:37.126568       1 main.go:56] Starting submariner-route-agent using the event framework
W0124 15:50:37.127711       1 client_config.go:608] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I0124 15:50:37.148017       1 cni_iface.go:75] Interface "lo" has "127.0.0.1" address
I0124 15:50:37.148061       1 cni_iface.go:75] Interface "lo" has "172.20.0.1" address
I0124 15:50:37.148067       1 cni_iface.go:75] Interface "lo" has "172.20.0.2" address
I0124 15:50:37.148072       1 cni_iface.go:75] Interface "lo" has "172.20.0.3" address
I0124 15:50:37.148077       1 cni_iface.go:75] Interface "lo" has "172.20.0.4" address
I0124 15:50:37.148082       1 cni_iface.go:75] Interface "lo" has "172.20.0.5" address
I0124 15:50:37.148087       1 cni_iface.go:75] Interface "lo" has "172.20.0.6" address
I0124 15:50:37.148092       1 cni_iface.go:75] Interface "lo" has "172.20.0.7" address
I0124 15:50:37.148097       1 cni_iface.go:75] Interface "lo" has "172.20.0.9" address
I0124 15:50:37.148102       1 cni_iface.go:75] Interface "lo" has "172.20.0.10" address
I0124 15:50:37.148106       1 cni_iface.go:75] Interface "lo" has "172.20.0.11" address
I0124 15:50:37.148259       1 cni_iface.go:75] Interface "eth0" has "10.249.0.7" address
I0124 15:50:37.148581       1 cni_iface.go:75] Interface "vethlocal" has "127.0.0.10" address
I0124 15:50:37.148881       1 cni_iface.go:75] Interface "tunl0" has "172.17.38.128" address
I0124 15:50:37.148902       1 cni_iface.go:80] Found CNI Interface "tunl0" that has IP "172.17.38.128" from ClusterCIDR "172.17.0.0/18"
I0124 15:50:37.229083       1 cni_iface.go:117] Successfully annotated node "10.249.0.7" with cniIfaceIP "172.17.38.128"
I0124 15:50:37.278369       1 registry.go:68] Event handler "logger" added to registry "routeagent_driver".
I0124 15:50:37.278392       1 registry.go:70] Event handler "kubeproxy-iptables-handler" ignored for registry "routeagent_driver".
I0124 15:50:37.278398       1 registry.go:70] Event handler "ovn-hostroutes-handler" ignored for registry "routeagent_driver".
I0124 15:50:37.278404       1 registry.go:68] Event handler "xfrm" added to registry "routeagent_driver".
I0124 15:50:37.278408       1 registry.go:68] Event handler "VXLAN cleanup handler" added to registry "routeagent_driver".
I0124 15:50:37.322539       1 ipset.go:351] Running ipset [create SUBMARINER-REMOTECIDRS hash:net -exist]
I0124 15:50:37.331554       1 ipset.go:351] Running ipset [create SUBMARINER-LOCALCIDRS hash:net -exist]
I0124 15:50:37.355425       1 iptables.go:114] In mangle table, iptables rule "-j SUBMARINER-POSTROUTING", exists at index 1.
I0124 15:50:37.355458       1 iptables.go:135] In mangle table, iptables rule "-j SUBMARINER-POSTROUTING", already exists.
I0124 15:50:37.393762       1 registry.go:68] Event handler "MTU handler" added to registry "routeagent_driver".
W0124 15:50:37.393936       1 client_config.go:608] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I0124 15:50:37.453600       1 controller.go:142] Starting the Event controller...
I0124 15:50:37.554767       1 handler.go:72] A new Endpoint for remote cluster "local-cluster" has been created: v1.EndpointSpec{ClusterID:"local-cluster", CableName:"submariner-cable-local-cluster-192-168-100-113", HealthCheckIP:"10.131.0.1", Hostname:"rdr-acm-submariner-48-tor01-worker-0", Subnets:[]string{"172.30.0.0/16", "10.128.0.0/14"}, PrivateIP:"192.168.100.113", PublicIP:"169.48.19.68", NATEnabled:false, Backend:"libreswan", BackendConfig:map[string]string{"natt-discovery-port":"4490", "preferred-server":"false", "udp-port":"4500"}}
I0124 15:50:37.554853       1 ipset.go:351] Running ipset [add SUBMARINER-REMOTECIDRS 172.30.0.0/16 -exist]
I0124 15:50:37.557963       1 ipset.go:351] Running ipset [add SUBMARINER-REMOTECIDRS 10.128.0.0/14 -exist]
I0124 15:50:37.559496       1 handler.go:57] A new Endpoint for the local cluster has been created: v1.EndpointSpec{ClusterID:"roks-48", CableName:"submariner-cable-roks-48-10-249-0-5", HealthCheckIP:"172.17.11.192", Hostname:"kube-c6onhkir0oii5kc8k6c0-rhacmsubmar-default-000003f8", Subnets:[]string{"172.21.0.0/16", "172.17.0.0/18"}, PrivateIP:"10.249.0.5", PublicIP:"163.66.87.59", NATEnabled:false, Backend:"libreswan", BackendConfig:map[string]string{"natt-discovery-port":"4490", "preferred-server":"false", "udp-port":"4500"}}
I0124 15:50:37.559536       1 ipset.go:351] Running ipset [add SUBMARINER-LOCALCIDRS 172.21.0.0/16 -exist]
I0124 15:50:37.560832       1 ipset.go:351] Running ipset [add SUBMARINER-LOCALCIDRS 172.17.0.0/18 -exist]
I0124 15:50:37.656391       1 controller.go:149] Event controller started
I0124 15:50:37.660687       1 handler.go:90] A Node with name "10.249.0.7" and addresses []v1.NodeAddress{v1.NodeAddress{Type:"InternalIP", Address:"10.249.0.7"}, v1.NodeAddress{Type:"ExternalIP", Address:"10.249.0.7"}, v1.NodeAddress{Type:"Hostname", Address:"10.249.0.7"}} has been added to the cluster
I0124 15:50:37.661119       1 handler.go:90] A Node with name "10.249.0.5" and addresses []v1.NodeAddress{v1.NodeAddress{Type:"InternalIP", Address:"10.249.0.5"}, v1.NodeAddress{Type:"ExternalIP", Address:"10.249.0.5"}, v1.NodeAddress{Type:"Hostname", Address:"10.249.0.5"}} has been added to the cluster
I0124 15:50:37.661732       1 handler.go:90] A Node with name "10.249.0.6" and addresses []v1.NodeAddress{v1.NodeAddress{Type:"InternalIP", Address:"10.249.0.6"}, v1.NodeAddress{Type:"ExternalIP", Address:"10.249.0.6"}, v1.NodeAddress{Type:"Hostname", Address:"10.249.0.6"}} has been added to the cluster
```

As you can see, these lines (when the fix is installed) are missing from the logs:
```
I0106 15:41:39.288529       1 iptables.go:31] Install/ensure SUBMARINER-POSTROUTING chain exists
I0106 15:41:39.339961       1 iptables.go:37] Insert SUBMARINER-POSTROUTING rule that has rules for inter-cluster traffic
I0106 15:41:39.364038       1 iptables.go:114] In nat table, iptables rule "-j SUBMARINER-POSTROUTING", exists at index 1.
I0106 15:41:39.364124       1 iptables.go:135] In nat table, iptables rule "-j SUBMARINER-POSTROUTING", already exists.
I0106 15:41:39.364143       1 iptables_iface.go:44] Install/ensure SUBMARINER-INPUT chain exists
I0106 15:41:39.409971       1 iptables_iface.go:55] Allow VxLAN incoming traffic in SUBMARINER-INPUT Chain
I0106 15:41:39.459885       1 iptables_iface.go:63] Insert rule to allow traffic over vx-submariner interface in FORWARDing Chain
I0106 15:41:39.492972       1 iptables_iface.go:75] Installing rule for host network to remote cluster communication: -s 240.0.0.0/8 -o vx-submariner -j SNAT --to 172.17.11.192
I0106 15:41:39.512978       1 registry.go:68] Event handler "kubeproxy-iptables-handler" added to registry "routeagent_driver".
```

**Comment by yboaron** [comment]
Thanks for the feedback,

So, from the logs you attached it seems that submariner-operator sets [1] the NetworkPlugin to 'Calico' after auto-discovery which means that submariner-operator image was built with the constant [2] NetworkPluginCalico = 'Calico'

Could it be that in the first case you used route-agent-image with the original constants.NetworkPluginCalico  value ('calico') ? 


[1]
https://github.com/submariner-io/submariner-operator/blob/fb8988c11a9521506926a3ea3f431a2422c7b63a/pkg/discovery/network/calico.go#L55
[2]
https://github.com/submariner-io/submariner/blob/27de30f905dc2da1b31d8e34c232f17fd1030752/pkg/routeagent_driver/constants/constants.go#L57

**Comment by dberg1** [comment]
@yboaron I did not rebuild the submariner-operator image. I only rebuilt the submariner-route-agent image with the 'Calico' constant and restarted the submariner-routeagent pods.
Notice that my cluster does not have any `calico-config` configmap.
It has a `cni-config` configmap in `calico-system` namespace:
```
# oc get configmap -A | grep calico
calico-system                                      cni-config                                                                   1      46d
calico-system                                      kube-root-ca.crt                                                             1      46d
calico-system                                      openshift-service-ca.crt                                                     1      46d
calico-system                                      typha-ca                                                                     1      46d
```

**Comment by dberg1** [comment]
So maybe a better fix would be to change the way the calico plugin is discovered, i.e. rely on something else than the existence of a `calico-config` configmap.

**Comment by yboaron** [comment]
From what I can see in sources, the flow is : 

A. Only the operator discovers network plugin and sets:
    - the right value in submariner_cr.status.NetworkPlugin 
    - SUBMARINER_NETWORKPLUGIN env variable  (in route agent DS) value to submariner_cr.status.NetworkPlugin
    
B. The route agent only activates the relevant handlers based on SUBMARINER_NETWORKPLUGIN value.

Hmm, I don't really understand how SUBMARINER_NETWORKPLUGIN env variable set to 'Calico' in your case if you didn't update the operator image, I would expect it to be set as 'unknown' plugin (since you don't have the calico-config configmap).

Bottom line we should understand/address two issues here:
1. How the operator should discover the calico plugin ( as you don't see the calico-config configmap)?
2. Why  SUBMARINER_NETWORKPLUGIN env variable in routeagent DS set to 'Calico' ?
 
For '2' I assume that submariner-operator  pod logs may help, could you please share the relevant logs?

**Comment by dberg1** [comment]
@yboaron I think it gets set to "Calico" by `discoverOpenShift4Network` in https://github.com/submariner-io/submariner-operator/blob/fb8988c11a9521506926a3ea3f431a2422c7b63a/pkg/discovery/network/network.go#L110
and
https://github.com/submariner-io/submariner-operator/blob/fb8988c11a9521506926a3ea3f431a2422c7b63a/pkg/discovery/network/openshift4.go#L93
based on:
```
apiVersion: config.openshift.io/v1
kind: Network
metadata:
  name: cluster
spec:
  clusterNetwork:
    - cidr: 172.17.0.0/18
      hostPrefix: 23
  externalIP:
    policy: {}
  networkType: Calico
  serviceNetwork:
    - 172.21.0.0/16
```
`discoverCalicoNetwork` is never called with OpenShift.

**Comment by dberg1** [comment]
So for OpenShift, we still need to recognize "Calico" as "calico" because `discoverCalicoNetwork` is not actually called.

**Comment by yboaron** [comment]
Haa, nice catch!
In that case, I think we should handle that in the operator auto-discovery code.

**Comment by yboaron** [comment]
https://github.com/submariner-io/submariner-operator/pull/1802 should address this issue:

**Comment by wdrdres3qew5ts21** [comment]
I see so in next version which is v0.12.0 will be fixed this problem right ? Currently I install from latest Bash script from your website it will be `v0.12.0` in my IBM Cloud with ROKS also facing same problem too.
![image](https://user-images.githubusercontent.com/25000903/151766609-e918c958-f6a5-421d-b862-b24df3baf581.png)
